pydra/engine/specs.py

```diff
@@ -90,23 +101,66 @@ def hash(self):
         """Compute a basic hash for any given set of fields."""
         from .helpers import hash_value, hash_function
 
+        # if inp_hash already calculated and nothing has changed, the old value can be used
+        if self.changed is False and self.inp_hash:
+            # if there are files in self.files_hash, checking if they didn't changed
+            check_files = [
+                Path(val[0]).stat().st_mtime == val[1]
+                for val in self.files_hash.values()
+            ]
+            if all(check_files):
+                return self.inp_hash
+        if self.files_hash is None:
+            self.files_hash = {}
         inp_dict = {}
         for field in attr_fields(self):
-            if field.name in ["_graph_checksums", "bindings"] or field.metadata.get(
-                "output_file_template"
-            ):
+            if field.name in [
+                "_graph_checksums",
+                "bindings",
+                "inp_hash",
+                "changed",
+                "files_hash",
+            ] or field.metadata.get("output_file_template"):
                 continue
             # removing values that are notset from hash calculation
             if getattr(self, field.name) is attr.NOTHING:
                 continue
+            value = getattr(self, field.name)
+            # checking if the field name is in the dictionary with pre-computed hash values for files
+            if field.name in self.files_hash:
+                filename, mtime, cont_hash = self.files_hash[field.name]
+                # if the name of the file and the time of last modification is the same,
+                # we are reusing the content hash value
+                if (
+                    str(Path(value)) == filename
+                    and Path(value).stat().st_mtime == mtime
+                ):
+                    inp_dict[field.name] = cont_hash
+                    continue
+
             inp_dict[field.name] = hash_value(
-                value=getattr(self, field.name), tp=field.type, metadata=field.metadata
+                value=value, tp=field.type, metadata=field.metadata
             )
+            # if file/dir, the hash value will be saved to prevent recomputation
+            if (
+                field.type in [File, Directory]
+                or "pydra.engine.specs.File" in str(field.type)
+                or "pydra.engine.specs.File" in str(field.type)
+            ) and is_existing_file(value):
```

Comment:
what if value is a list of files or some other structure with File as an inner type. this was taken care of in hash_value

Reply:
in the new version i've changed this completely - `hash_file` compares the modification times and hashes